src/librustc/mir/interpret/error.rs

```diff
@@ -309,8 +311,46 @@ impl<O: fmt::Debug> fmt::Debug for PanicMessage<O> {
     }
 }
 
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum InvalidProgramMessage {
+    /// Resolution can fail if we are in a too generic context
+    TooGeneric,
+    /// Cannot compute this constant because it depends on another one
+    /// which already produced an error
+    ReferencedConstant,
+    /// Abort in case type errors are reached
+    TypeckError,
+}
+
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum UndefinedBehaviourMessage {
+}
+
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum UnsupportedMessage {
+}
+
+#[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
+pub enum ResourceExhaustionMessage {
+}
+
 #[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
 pub enum InterpError<'tcx> {
+    /// The program panicked.
+    Panic(PanicMessage<u64>),
+    /// The program caused undefined behavior.
+    UndefinedBehaviour(UndefinedBehaviourMessage),
+    /// The program did something the interpreter does not support (some of these *might* be UB
+    /// but the interpreter is not sure).
+    Unsupported(UnsupportedMessage),
```

Comment:
In its this type looks strange now. Maybe `UnsupportedOperation(UnsupportedOperationMessage)`? That's very long though.

Reply:
Or what about `UnsupportedOpInfo`? That's not terribly long.

`UnsupportedInfo` sounds a bit like the information is unsupported.

Reply:
@saleemjaffer reminder that this comment is still open. :)

There are also some more open comments below; you can see them all by scrolling through the PR page.

Reply:
Cool, I'll use `UnsupportedOpInfo`.

Reply:
Still looks like `UnsupportedInfo` in the latest diff?